src/analyzer.ts

```diff
@@ -82,102 +90,117 @@ export class Analyzer {
 
   /**
    * Loads, parses and analyzes a document and its transitive dependencies.
-   *
-   * @param {string} url the location of the file to analyze
-   * @return {Promise<DocumentDescriptor>}
    */
-  async analyze(url: string): Promise<DocumentDescriptor> {
-    return this._analyzeResolved(this._resolveUrl(url));
+  async analyze(url: string): Promise<Document|null> {
+    const scannedDocument = await this._analyzeResolved(this._resolveUrl(url));
+    return Document.makeRootDocument(scannedDocument);
   }
 
-  private async _analyzeResolved(resolvedUrl: string):
-      Promise<DocumentDescriptor> {
-    let isExternalUrl = urlLib.parse(resolvedUrl).protocol !== null ||
-        resolvedUrl.startsWith('//');
-    if (isExternalUrl) {
-      return null;
-    }
-    let cachedResult = this._documentDescriptors.get(resolvedUrl);
+  /**
+   * Like `analyze`, but for use cases like editors where the caller believes
+   * the file has changed since the last time it was analyzed, and you may
+   * even have unsaved in-memory contents to use for the given file.
+   */
+  async analyzeChangedFile(url: string, contents?: string):
+      Promise<Document|null> {
+    const resolved = this._resolveUrl(url);
+    this._scannedDocuments.delete(resolved);
+    this._parsedDocuments.delete(resolved);
+    const scannedDocument =
+        await this._analyzeResolved(this._resolveUrl(url), contents);
+    return Document.makeRootDocument(scannedDocument);
+  }
+
+  private async _analyzeResolved(resolvedUrl: string, contents?: string):
+      Promise<ScannedDocument|null> {
+    let cachedResult = this._scannedDocuments.get(resolvedUrl);
     if (cachedResult) {
       return cachedResult;
     }
     let promise = (async() => {
       // Make sure we wait and return a Promise before doing any work, so that
-      // the Promise can be cached.
+      // the Promise is cached before anything else happens.
       await Promise.resolve();
-      let document = await this._loadResolved(resolvedUrl);
+      let document = await this._loadResolved(resolvedUrl, contents);
       return this._analyzeDocument(document);
     })();
-    this._documentDescriptors.set(resolvedUrl, promise);
+    this._scannedDocuments.set(resolvedUrl, promise);
     return promise;
   }
 
-  async resolve(): Promise<Analysis> {
-    return new Analysis(await Promise.all(this._documentDescriptors.values()));
+  private async _resolve(url: string): Promise<ScannedDocument|null> {
```

Comment:
also wondering when this is null. We should decide/comment what's an error case vs null.